core/src/impl/Kokkos_Core.cpp

```diff
@@ -282,11 +295,206 @@ void initialize_profiling(const InitArguments& args) {
     std::exit(EXIT_SUCCESS);
   }
   Kokkos::Tools::parseArgs(args.tool_args);
+  for (const auto& category_values : Kokkos::Impl::metadata_map) {
+    for (const auto& key_value : category_values.second) {
+      Kokkos::Tools::declareMetadata(key_value.first, key_value.second);
+    }
+  }
 }
-
+#define KOKKOS_METADATA_QUOTE_HELPER(x) #x
+#define KOKKOS_METADATA_QUOTE(x) KOKKOS_METADATA_QUOTE_HELPER(x)
 void pre_initialize_internal(const InitArguments& args) {
   if (args.disable_warnings) g_show_warnings = false;
   if (args.tune_internals) g_tune_internals = true;
+  std::stringstream str_builder;
+  str_builder << KOKKOS_VERSION / 10000 << "." << (KOKKOS_VERSION % 10000) / 100
+              << "." << KOKKOS_VERSION % 100;
+  declare_configuration_metadata("version_info", "Kokkos Version",
+                                 str_builder.str());
+  declare_configuration_metadata("version_info", "Kokkos SHA",
+                                 KOKKOS_METADATA_QUOTE(KOKKOS_GIT_REVISION));
+#ifdef KOKKOS_COMPILER_APPLECC
+  declare_configuration_metadata("compiler_version", "KOKKOS_COMPILER_APPLECC",
+                                 std::to_string(KOKKOS_COMPILER_APPLECC));
+  declare_configuration_metadata("tools_only", "compiler_family", "apple");
+#endif
+#ifdef KOKKOS_COMPILER_CLANG
+  declare_configuration_metadata("compiler_version", "KOKKOS_COMPILER_CLANG",
+                                 std::to_string(KOKKOS_COMPILER_CLANG));
+  declare_configuration_metadata("tools_only", "compiler_family", "clang");
+#endif
+#ifdef KOKKOS_COMPILER_CRAYC
+  declare_configuration_metadata("compiler_version", "KOKKOS_COMPILER_CRAYC",
+                                 std::to_string(KOKKOS_COMPILER_CRAYC));
+  declare_configuration_metadata("tools_only", "compiler_family", "cray");
+#endif
+#ifdef KOKKOS_COMPILER_GNU
+  declare_configuration_metadata("compiler_version", "KOKKOS_COMPILER_GNU",
+                                 std::to_string(KOKKOS_COMPILER_GNU));
+  declare_configuration_metadata("tools_only", "compiler_family", "gnu");
+#endif
+#ifdef KOKKOS_COMPILER_IBM
+  declare_configuration_metadata("compiler_version", "KOKKOS_COMPILER_IBM",
+                                 std::to_string(KOKKOS_COMPILER_IBM));
+  declare_configuration_metadata("tools_only", "compiler_family", "ibm");
+#endif
+#ifdef KOKKOS_COMPILER_INTEL
+  declare_configuration_metadata("compiler_version", "KOKKOS_COMPILER_INTEL",
+                                 std::to_string(KOKKOS_COMPILER_INTEL));
+  declare_configuration_metadata("tools_only", "compiler_family", "intel");
+#endif
```

Comment:
What about MSVC?

Reply:
I thought this was just a copy of the old version. If we want to add MSVC, I'd prefer it be in another PR

Reply:
Hmmm... I get
```bash
$ git grep "#define KOKKOS_COMPILER" | sort -u 
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_APPLECC __APPLE_CC__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_CLANG \
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_CRAYC _CRAYC
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_GNU \
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_IBM __IBMC__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_IBM __IBMCPP__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_IBM __ibmxl_vrm__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_INTEL __ECC
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_INTEL __ICC
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_INTEL __INTEL_COMPILER
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_INTEL __INTEL_LLVM_COMPILER
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_MSVC _MSC_VER
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_NVCC __NVCC__
core/src/Kokkos_Macros.hpp:#define KOKKOS_COMPILER_PGI \
```
so `MSVCC` really is the only one missing here and I would prefer to add it in this pull request.

Reply:
Do we want to add it to print_configuration, though? In the current implementation, print_configuration in Core doesn't say anything about MSVC. So we have three options

1) Change and add it to the printed metadata
2) Only add it to the declared metadata
3) Don't add it at all

I have no preferences among the three, it sounds like you prefer (1), but just to check, would you prefer (1), or (2)

Reply:
Talked in Slack, implemented (1) via 3e95a906